app/src/main/java/org/oppia/app/player/state/itemviewmodel/RatioExpressionInputInteractionViewModel.kt

```diff
@@ -0,0 +1,113 @@
+
+import android.content.Context
+import android.text.Editable
+import android.text.TextWatcher
+import androidx.databinding.Observable
+import androidx.databinding.ObservableField
+import org.oppia.app.R
+import org.oppia.app.model.Interaction
+import org.oppia.app.model.InteractionObject
+import org.oppia.app.model.UserAnswer
+import org.oppia.app.parser.StringToRatioParser
+import org.oppia.app.player.state.answerhandling.AnswerErrorCategory
+import org.oppia.app.player.state.answerhandling.InteractionAnswerErrorOrAvailabilityCheckReceiver
+import org.oppia.app.player.state.answerhandling.InteractionAnswerHandler
+import org.oppia.app.utility.toAccessibleAnswerString
+import org.oppia.domain.util.toAnswerString
+
+/** [StateItemViewModel] for the ratio expression input interaction. */
+class RatioExpressionInputInteractionViewModel(
+  interaction: Interaction,
+  private val context: Context,
+  val hasConversationView: Boolean,
+  val isSplitView: Boolean,
+  private val errorOrAvailabilityCheckReceiver: InteractionAnswerErrorOrAvailabilityCheckReceiver
+) : StateItemViewModel(ViewType.RATIO_EXPRESSION_INPUT_INTERACTION), InteractionAnswerHandler {
+  private var pendingAnswerError: String? = null
+  var answerText: CharSequence = ""
+  var isAnswerAvailable = ObservableField<Boolean>(false)
+  var errorMessage = ObservableField<String>("")
+
+  val hintText: CharSequence = deriveHintText(interaction)
+  private val stringToRatioParser: StringToRatioParser = StringToRatioParser()
+  private val numberOfTerms =
+    interaction.customizationArgsMap["numberOfTerms"]?.signedInt ?: 0
+
+  init {
+    val callback: Observable.OnPropertyChangedCallback =
+      object : Observable.OnPropertyChangedCallback() {
+        override fun onPropertyChanged(sender: Observable, propertyId: Int) {
+          errorOrAvailabilityCheckReceiver.onPendingAnswerErrorOrAvailabilityCheck(
+            pendingAnswerError,
+            answerText.isNotEmpty()
+          )
+        }
+      }
+    errorMessage.addOnPropertyChangedCallback(callback)
+    isAnswerAvailable.addOnPropertyChangedCallback(callback)
+  }
+
+  override fun getPendingAnswer(): UserAnswer {
+    val userAnswerBuilder = UserAnswer.newBuilder()
+    if (answerText.isNotEmpty()) {
+      val ratioAnswer = stringToRatioParser.parseRatioOrThrow(answerText.toString())
+      userAnswerBuilder.answer = InteractionObject.newBuilder()
+        .setRatioExpression(ratioAnswer)
+        .build()
+      userAnswerBuilder.plainAnswer = ratioAnswer.toAnswerString()
+      userAnswerBuilder.contentDescription = ratioAnswer.toAccessibleAnswerString(context)
+    }
+    return userAnswerBuilder.build()
+  }
+
+  /** It checks the pending error for the current ratio input, and correspondingly updates the error string based on the specified error category. */
+  override fun checkPendingAnswerError(category: AnswerErrorCategory): String? {
+    if (answerText.isNotEmpty()) {
+      when (category) {
```

Comment:
Do you need to reset ``pendingAnswerError`` to null in the 'else' case?

Reply:
why do I need to do that?

Reply:
Otherwise ``pendingAnswerError`` has an error value even in cases when there's no actual error.

Reply:
but if we are using an `enum` class with `when` else condition won't be executed

Reply:
Ah I see what you mean. ``pendingAnswerError`` should reset correctly based on the returned value being VALID. Thanks for clarifying.